app/controllers/AddressController.scala

```diff
@@ -331,80 +332,111 @@ class AddressController @Inject()(
                              buildEvent(
                                "addressLookupResults",
                                "find_address",
-                               Map("postcode" -> Some(addressFinderDto.postcode), "filter" -> addressFinderDto.filter)))
-                           Future.successful(
-                             Ok(
-                               views.html.personaldetails.addressSelector(
-                                 AddressSelectorDto.form,
-                                 recordSet,
-                                 typ,
-                                 addressFinderDto.postcode,
-                                 addressFinderDto.filter
-                               )
-                             )
-                           )
+                               Map(POSTCODE -> Some(addressFinderDto.postcode), FILTER -> addressFinderDto.filter)))
+
+                           cacheSelectedRecordSet(typ, recordSet) map { _ =>
+                             Redirect(routes.AddressController.showAddressSelectorForm(typ))
+                               .addingToSession(
+                                 (POSTCODE, addressFinderDto.postcode),
+                                 (FILTER, addressFinderDto.filter.getOrElse("")))
+                           }
                        }
             } yield result
           }
         )
       }
     }
 
+  def showAddressSelectorForm(typ: AddrType) =
+    authenticate.async { implicit request =>
+      val postcode =
+        request.body.asFormUrlEncoded.flatMap(_.get(POSTCODE).flatMap(_.headOption)).getOrElse("")
+      val filter =
+        request.body.asFormUrlEncoded.flatMap(_.get(FILTER).flatMap(_.headOption))
+
+      gettingCachedJourneyData(typ) { journeyData =>
+        journeyData.recordSet match {
+          case Some(set) =>
+            Future.successful(
+              Ok(
+                views.html.personaldetails.addressSelector(
+                  AddressSelectorDto.form,
+                  set,
+                  typ,
+                  postcode,
+                  filter
+                )
+              )
+            )
+          case _ => Future.successful(Redirect(routes.AddressController.showPostcodeLookupForm(typ)))
+        }
+      }
+    }
+
   def processAddressSelectorForm(typ: AddrType): Action[AnyContent] =
     authenticate.async { implicit request =>
       val postcode =
-        request.body.asFormUrlEncoded.flatMap(_.get("postcode").flatMap(_.headOption)).getOrElse("")
+        request.body.asFormUrlEncoded.flatMap(_.get(POSTCODE).flatMap(_.headOption)).getOrElse("")
       val filter =
-        request.body.asFormUrlEncoded.flatMap(_.get("filter").flatMap(_.headOption))
+        request.body.asFormUrlEncoded.flatMap(_.get(FILTER).flatMap(_.headOption))
+
+      val errorPage = Future.successful(
```

Comment:
This same code is used in  a few places in this file, pull it out into it's own method